### ISSUE TITLE
debug the golang project use root privilege

### ISSUE THREAD

**Issue by calmkart** [microsoft/vscode-go#2889]
**Is your feature request related to a problem? Please describe.**

i want to debug a golang project in vscode, but the project need root privilege(i can run my project "sudo go main.go").the vscode debug launch.json for golang has no privilege settings.i try to find a way to debug in root, but there is not a easy method.

**Describe the solution you'd like**

add a root privilege settings option in golang launch.json like [python debug](https://donjayamanne.github.io/pythonVSCodeDocs/docs/debugging_sudo-root-privileges/).

**Describe alternatives you've considered**

run vscode in root privilege(sudo run vscode), but i don't think it's a good solution.

**Additional context**

None

**Comment by ramya-rao-a** [comment]
@jhendrixMSFT, @quoctruong thoughts?

**Comment by quoctruong** [comment]
@calmkart Another alternative you can consider is to launch the program with `sudo` and attach to it.

**Comment by fastcat** [comment]
Expanding on what @quoctruong suggested ...

I got a tip from somewhere else on a way to do this, which I adapted for a project of mine: Use a wrapper script to invoke `dlv` with `sudo`, and set that under `go.alternateTools`.

For example, in a repo of mine I have:
`.vscode/settings.json`:
```json
{
	"go.alternateTools": {
		"dlv": "${workspaceFolder}/.vscode/dlv-sudo.sh"
	}
}
```
and `.vscode/dlv-sudo.sh`:
```shell
#!/bin/sh
if ! which dlv ; then
	PATH="${GOPATH}/bin:$PATH"
fi
if [ "$DEBUG_AS_ROOT" = "true" ]; then
	exec sudo dlv "$@"
else
	exec dlv "$@"
fi
```
(don't forget to `chmod +x` this script)
I check the environment variable because I have some targets in my project that don't need (and thus shouldn't have) root, and others that do.  YMMV.

And finally, for the launch task(s) that need root:
`.vscode/launch.json`:
```jsonc
{
	"version": "0.2.0",
	"configurations": [
		{
			"name": "Run as root",
			"type": "go",
			"request": "launch",
			"mode": "auto",
			"program": "${workspaceFolder}",
			"env": {
				"DEBUG_AS_ROOT": "true",
			}
			// etc...
		}
	]
}
```

**Comment by stamblerre** [comment]
@quoctruong: Does this issue still need further discussion or is this working as intended?

**Comment by quoctruong** [comment]
It's working as intended. Adding `sudo` would be a nice to have but I don't think that's necessary.

**Comment by stamblerre** [comment]
Makes sense. https://github.com/microsoft/vscode-go/issues/2610 is a duplicate that mentions  debugging with root privileges remotely - is there a work-around for that?

**Comment by quoctruong** [comment]
I think the work around by starting the remote dlv session with `sudo` should still work.

**Comment by stamblerre** [comment]
Sounds good. @ramya-rao-a: Should this issue be closed or left open for others to find?

**Comment by quoctruong** [comment]
@ramya-rao-a I think this issue can be closed since the workaround is available and we don't want to add extra field to the current debug configuration.

**Comment by TheDiveO** [comment]
Unfortunately, the above mentioned remedy doesn't work out of the box any more, or it lacks some sudo configuration I'm not aware of. Also, running dlv as sudo without `--only-same-user=false` will fail and debugging seems to be hung. So these are the pieces to actually get it working:

1. Update `dlv` to the master branch in order to get the fix for [dlv issue #1835](https://github.com/go-delve/delve/issues/1835): `https://github.com/go-delve/delve/issues/1835@master`
2. Edit `/etc/sudoers` using `visudo` and append(!) the following last line to enable password-less sudo to `dlv`; make sure to set the correct user and path:
   ```
   userfoo ALL=(root)NOPASSWD:/home/userfoo/go/bin/dlv
   ```
   You need to append this line, as sudo takes the last matching line, so if you have a more general rule for your user last, then that will match -- something we don't want in this situation.
3. Create a file `.vscode/dlv-sudo.sh`, with some important changes from the version shown above:
   ```bash
   #!/bin/sh
   if ! which dlv ; then
   	PATH="${GOPATH}/bin:$PATH"
   fi
   if [ "$DEBUG_AS_ROOT" = "true" ]; then
   	DLV=$(which dlv)
   	exec sudo "$DLV" --only-same-user=false "$@"
   else
   	exec dlv "$@"
   fi
   ```
   The important change is to use the full path to `dlv` to make the password-less sudo work, and to pass `--only-same-user=false` to `dlv` to allow vsc under your user to connect to root's `dlv`.
4. Create a suitable `launch.json`, such as:
   ```json
   {
       "version": "0.2.0",
       "configurations": [
           {
               "name": "Test as root",
               "type": "go",
               "request": "launch",
               "mode": "test",
               "program": "${fileDirname}",
               "env": {
                   "DEBUG_AS_ROOT": "true",
               },
           },
       ],
       "compounds": []
    }
   ```
5. Now debug and test sudo golang programs and packages to your pleasure.

**Comment by ramya-rao-a** [comment]
@TheDiveO If you are still seeing problems, please log a new issue at https://github.com/golang/vscode-go as [we are moving](https://github.com/microsoft/vscode-go#we-are-moving---)
